Replace WebCore::notifyHistoryItemChanged with a proper client callback

achristensen07 opened the pull request:
#### def02cb5c4c24ed8a05e4e5075eb6ec95216ef80
<pre>
Replace WebCore::notifyHistoryItemChanged with a proper client callback
<a href="https://bugs.webkit.org/show_bug.cgi?id=261223">https://bugs.webkit.org/show_bug.cgi?id=261223</a>
rdar://115072660

Reviewed by Chris Dumez.

Using a process-global function pointer with little context doesn&apos;t allow us to manage state and scope very well.
A client object does.

* Source/WebCore/history/HistoryItem.cpp:
(WebCore::HistoryItem::HistoryItem):
(WebCore::m_client):
(WebCore::HistoryItem::notifyChanged):
(WebCore::defaultNotifyHistoryItemChanged): Deleted.
(WebCore::m_identifier): Deleted.
* Source/WebCore/history/HistoryItem.h:
(WebCore::HistoryItem::create):
* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::pageConfigurationWithEmptyClients):
* Source/WebCore/loader/HistoryController.cpp:
(WebCore::FrameLoader::HistoryController::updateForRedirectWithLockedBackForwardList):
(WebCore::FrameLoader::HistoryController::createItem):
(WebCore::FrameLoader::HistoryController::createItemTree):
(WebCore::FrameLoader::HistoryController::updateBackForwardListClippedAtTarget):
(WebCore::FrameLoader::HistoryController::pushState):
* Source/WebCore/loader/HistoryController.h:
* Source/WebCore/page/Page.cpp:
(WebCore::m_historyItemClient):
(WebCore::m_badgeClient): Deleted.
* Source/WebCore/page/Page.h:
(WebCore::Page::historyItemClient):
* Source/WebCore/page/PageConfiguration.cpp:
(WebCore::PageConfiguration::PageConfiguration):
* Source/WebCore/page/PageConfiguration.h:
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebCoreSupport/SessionStateConversion.cpp:
(WebKit::applyFrameState):
(WebKit::toHistoryItem):
* Source/WebKit/WebProcess/WebCoreSupport/SessionStateConversion.h:
* Source/WebKit/WebProcess/WebPage/HistoryItemClient.cpp: Copied from Source/WebKit/WebProcess/WebCoreSupport/SessionStateConversion.h.
(WebKit::HistoryItemClient::ignoreChangesForScope):
(WebKit::HistoryItemClient::historyItemChanged):
* Source/WebKit/WebProcess/WebPage/HistoryItemClient.h: Copied from Source/WebKit/WebProcess/WebCoreSupport/SessionStateConversion.h.
(WebKit::HistoryItemClient::create):
* Source/WebKit/WebProcess/WebPage/MomentumEventDispatcher.cpp:
* Source/WebKit/WebProcess/WebPage/WebBackForwardListProxy.cpp:
(WebKit::WebBackForwardListProxy::WebBackForwardListProxy):
(WebKit::WK2NotifyHistoryItemChanged): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_historyItemClient):
(WebKit::WebPage::restoreSessionInternal):
(WebKit::WebPage::setCurrentHistoryItemForReattach):
(WebKit::m_appHighlightsVisible): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKitLegacy/mac/History/WebHistoryItem.mm:
(WKNotifyHistoryItemChanged):
(GlobalHistoryItemClient::create):
(GlobalHistoryItemClient::historyItemChanged):
(globalHistoryItemClient):
(-[WebHistoryItem init]):
(-[WebHistoryItem initWithURLString:title:lastVisitedTimeInterval:]):
(-[WebHistoryItem initWithURLString:title:displayTitle:lastVisitedTimeInterval:]):
(-[WebHistoryItem initWithWebCoreHistoryItem:]):
* Source/WebKitLegacy/mac/History/WebHistoryItemInternal.h:
* Source/WebKitLegacy/mac/WebView/WebFrameView.mm:
(-[WebFrameView initWithFrame:]):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(LegacyHistoryItemClient::create):
(LegacyHistoryItemClient::historyItemChanged):
(-[WebView _commonInitializationWithFrameName:groupName:]):
(-[WebView initSimpleHTMLDocumentWithStyle:frame:preferences:groupName:]):

Canonical link: <a href="https://commits.webkit.org/267857@main">https://commits.webkit.org/267857@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/379145bd7db31b981365a99c120c4ef2f9c7492f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17759 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18089 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/18627 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/19583 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/16612 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17954 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/21380 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18239 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/18658 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17972 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18263 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/15431 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/20449 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/15495 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/16184 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/22755 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/16507 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/16352 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20616 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16920 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/14327 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16033 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/20396 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2195 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16765 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/loader/EmptyClients.cpp:
```diff
@@ -1187,6 +1187,13 @@ class EmptySocketProvider final : public SocketProvider {
 
+class EmptyHistoryItemClient final : public HistoryItemClient {
+public:
+    static Ref<EmptyHistoryItemClient> create() { return adoptRef(*new EmptyHistoryItemClient()); }
```
nit: `()` after `new EmptyHistoryItemClient` is not needed.

cdumez commented on Source/WebKit/WebProcess/WebPage/HistoryItemClient.h:
```diff
@@ -0,0 +1,45 @@
+
+namespace WebKit {
+
+class HistoryItemClient : public WebCore::HistoryItemClient {
```
Can we add 'Web' prefix to the name. I really hate it when classes have the same name in WebKit and WebCore. Makes it annoying to lookup a class by file name in the IDE.

Also, can we mark this class as `final`?

cdumez commented on Source/WebKit/WebProcess/WebPage/HistoryItemClient.h:
```diff
@@ -0,0 +1,45 @@
+private:
+    void historyItemChanged(const WebCore::HistoryItem&) final;
+
+    bool m_ignoreChanges { false };
```
Needs a prefix, e.g. `m_shouldIgnoreChanges`

cdumez commented on Source/WebKit/WebProcess/WebPage/HistoryItemClient.h:
```diff
@@ -0,0 +1,45 @@
+
+    ScopeExit<CompletionHandler<void()>> ignoreChangesForScope();
+
+private:
```
We should probably mark the constructor as private. The default constructor probably is public, which means someone could makeUnique<>().

cdumez commented on Source/WebKitLegacy/mac/History/WebHistoryItem.mm:
```diff
@@ -120,6 +120,22 @@ void WKNotifyHistoryItemChanged(HistoryItem&)
 #endif
 }
 
+class GlobalHistoryItemClient : public WebCore::HistoryItemClient {
```
Mark as `final` ?

cdumez commented on Source/WebKitLegacy/mac/History/WebHistoryItem.mm:
```diff
@@ -120,6 +120,22 @@ void WKNotifyHistoryItemChanged(HistoryItem&)
+class GlobalHistoryItemClient : public WebCore::HistoryItemClient {
+public:
+    static Ref<GlobalHistoryItemClient> create() { return adoptRef(*new GlobalHistoryItemClient()); }
+private:
```
Mark constructor as private?

cdumez commented on Source/WebKitLegacy/mac/History/WebHistoryItem.mm:
```diff
@@ -120,6 +120,22 @@ void WKNotifyHistoryItemChanged(HistoryItem&)
 
+class GlobalHistoryItemClient : public WebCore::HistoryItemClient {
+public:
+    static Ref<GlobalHistoryItemClient> create() { return adoptRef(*new GlobalHistoryItemClient()); }
```
`()` is unnecessary.

cdumez commented on Source/WebKit/WebProcess/WebPage/HistoryItemClient.h:
```diff
@@ -0,0 +1,45 @@
+
+class HistoryItemClient : public WebCore::HistoryItemClient {
+public:
+    static Ref<HistoryItemClient> create() { return adoptRef(*new HistoryItemClient()); }
```
`()` is unnecessary.

cdumez commented on Source/WebKitLegacy/mac/WebView/WebView.mm:
```diff
@@ -820,6 +820,16 @@ void tryDestruction()
 
 } // namespace WebKit
 
+class LegacyHistoryItemClient : public WebCore::HistoryItemClient {
```
I am confused as to why WebKit legacy needs both `GlobalHistoryItemClient` AND this class? What prevents us from using only one? How do they differ?

cdumez replied:
Not r+'ing yet because of this.

cdumez commented on Source/WebCore/history/HistoryItem.h:
```diff
@@ -57,30 +57,22 @@ class Image;
 enum class PruningReason;
 
-WEBCORE_EXPORT extern void (*notifyHistoryItemChanged)(HistoryItem&);
+class HistoryItemClient : public RefCounted<HistoryItemClient> {
```
Seems we could move this inside the `HistoryItem` class and rename to `Client`? At the very least, we could add a `using Client = HistoryItemClient;` inside `HistoryItem`.

cdumez commented on Source/WebKitLegacy/WebKitLegacy.xcodeproj/project.pbxproj:
```diff
@@ -3,7 +3,7 @@
 	classes = {
 	};
-	objectVersion = 52;
+	objectVersion = 54;
```
We should revert this part.

cdumez commented on Source/WebKitLegacy/mac/WebView/WebView.mm:
```diff
@@ -1520,6 +1521,7 @@ - (void)_commonInitializationWithFrameName:(NSString *)frameName groupName:(NSSt
         makeUniqueRef<WebCore::DummyStorageProvider>(),
         makeUniqueRef<WebCore::DummyModelPlayerProvider>(),
         WebCore::EmptyBadgeClient::create(),
+        LegacyHistoryItemClient::create(),
```
Do we really need this LegacyHistoryItemClient? Why can't we just use globalHistoryItemClient() consistently in WebKitLegacy?

achristensen07 replied:
I'll use a singleton.  They're kind of equivalent because it only calls a global function, but you're right a singleton would have fewer allocations with many WebViews.

achristensen07 committed:
Replace WebCore::notifyHistoryItemChanged with a proper client callback
https://bugs.webkit.org/show_bug.cgi?id=261223
rdar://115072660

Reviewed by Chris Dumez.

Using a process-global function pointer with little context doesn't allow us to manage state and scope very well.
A client object does.

* Source/WebCore/history/HistoryItem.cpp:
(WebCore::HistoryItem::HistoryItem):
(WebCore::m_client):
(WebCore::HistoryItem::notifyChanged):
(WebCore::defaultNotifyHistoryItemChanged): Deleted.
(WebCore::m_identifier): Deleted.
* Source/WebCore/history/HistoryItem.h:
(WebCore::HistoryItem::create):
* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::pageConfigurationWithEmptyClients):
* Source/WebCore/loader/HistoryController.cpp:
(WebCore::FrameLoader::HistoryController::updateForRedirectWithLockedBackForwardList):
(WebCore::FrameLoader::HistoryController::createItem):
(WebCore::FrameLoader::HistoryController::createItemTree):
(WebCore::FrameLoader::HistoryController::updateBackForwardListClippedAtTarget):
(WebCore::FrameLoader::HistoryController::pushState):
* Source/WebCore/loader/HistoryController.h:
* Source/WebCore/page/Page.cpp:
(WebCore::m_historyItemClient):
(WebCore::m_badgeClient): Deleted.
* Source/WebCore/page/Page.h:
(WebCore::Page::historyItemClient):
* Source/WebCore/page/PageConfiguration.cpp:
(WebCore::PageConfiguration::PageConfiguration):
* Source/WebCore/page/PageConfiguration.h:
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebCoreSupport/SessionStateConversion.cpp:
(WebKit::applyFrameState):
(WebKit::toHistoryItem):
* Source/WebKit/WebProcess/WebCoreSupport/SessionStateConversion.h:
* Source/WebKit/WebProcess/WebPage/HistoryItemClient.cpp: Copied from Source/WebKit/WebProcess/WebCoreSupport/SessionStateConversion.h.
(WebKit::HistoryItemClient::ignoreChangesForScope):
(WebKit::HistoryItemClient::historyItemChanged):
* Source/WebKit/WebProcess/WebPage/HistoryItemClient.h: Copied from Source/WebKit/WebProcess/WebCoreSupport/SessionStateConversion.h.
(WebKit::HistoryItemClient::create):
* Source/WebKit/WebProcess/WebPage/MomentumEventDispatcher.cpp:
* Source/WebKit/WebProcess/WebPage/WebBackForwardListProxy.cpp:
(WebKit::WebBackForwardListProxy::WebBackForwardListProxy):
(WebKit::WK2NotifyHistoryItemChanged): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_historyItemClient):
(WebKit::WebPage::restoreSessionInternal):
(WebKit::WebPage::setCurrentHistoryItemForReattach):
(WebKit::m_appHighlightsVisible): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKitLegacy/mac/History/WebHistoryItem.mm:
(WKNotifyHistoryItemChanged):
(GlobalHistoryItemClient::create):
(GlobalHistoryItemClient::historyItemChanged):
(globalHistoryItemClient):
(-[WebHistoryItem init]):
(-[WebHistoryItem initWithURLString:title:lastVisitedTimeInterval:]):
(-[WebHistoryItem initWithURLString:title:displayTitle:lastVisitedTimeInterval:]):
(-[WebHistoryItem initWithWebCoreHistoryItem:]):
* Source/WebKitLegacy/mac/History/WebHistoryItemInternal.h:
* Source/WebKitLegacy/mac/WebView/WebFrameView.mm:
(-[WebFrameView initWithFrame:]):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(LegacyHistoryItemClient::create):
(LegacyHistoryItemClient::historyItemChanged):
(-[WebView _commonInitializationWithFrameName:groupName:]):
(-[WebView initSimpleHTMLDocumentWithStyle:frame:preferences:groupName:]):

Canonical link: https://commits.webkit.org/267857@main
